packages/flutter/lib/src/gestures/mouse_tracking.dart

```diff
@@ -152,194 +245,212 @@ class MouseTracker extends ChangeNotifier {
     // Order is important for mouse event callbacks. The `findAnnotations`
     // returns annotations in the visual order from front to back. We call
     // it the "visual order", and the opposite one "reverse visual order".
     // The algorithm here is explained in
     // https://github.com/flutter/flutter/issues/41420
 
-    // The `nextAnnotations` is annotations that contains this device in the
-    // coming frame in visual order.
-    // Order is preserved with the help of [LinkedHashSet].
-
-    final PointerEvent mostRecentEvent = currentState.mostRecentEvent;
-    // The `lastAnnotations` is annotations that contains this device in the
-    // previous frame in visual order.
-    final LinkedHashSet<MouseTrackerAnnotation> lastAnnotations = currentState.lastAnnotations;
-
     // Send exit events in visual order.
     final Iterable<MouseTrackerAnnotation> exitingAnnotations =
       lastAnnotations.difference(nextAnnotations);
     for (final MouseTrackerAnnotation annotation in exitingAnnotations) {
-      if (annotation.onExit != null) {
-        annotation.onExit(PointerExitEvent.fromMouseEvent(mostRecentEvent));
+      final bool attached = isAnnotationAttached(annotation);
+      if (annotation.onExit != null && attached) {
+        annotation.onExit(PointerExitEvent.fromMouseEvent(latestEvent));
       }
     }
 
     // Send enter events in reverse visual order.
     final Iterable<MouseTrackerAnnotation> enteringAnnotations =
       nextAnnotations.difference(lastAnnotations).toList().reversed;
     for (final MouseTrackerAnnotation annotation in enteringAnnotations) {
+      assert(isAnnotationAttached(annotation));
       if (annotation.onEnter != null) {
-        annotation.onEnter(PointerEnterEvent.fromMouseEvent(mostRecentEvent));
+        annotation.onEnter(PointerEnterEvent.fromMouseEvent(latestEvent));
       }
     }
 
     // Send hover events in reverse visual order.
     // For now the order between the hover events is designed this way for no
     // solid reasons but to keep it aligned with enter events for simplicity.
-    if (mostRecentEvent is PointerHoverEvent) {
+    if (latestEvent is PointerHoverEvent) {
       final Iterable<MouseTrackerAnnotation> hoveringAnnotations =
         nextAnnotations.toList().reversed;
       for (final MouseTrackerAnnotation annotation in hoveringAnnotations) {
-        if (annotation.onHover != null) {
-          annotation.onHover(mostRecentEvent);
+        // Deduplicate: Trigger hover if it's a newly hovered annotation
+        // or the position has changed
+        if (!lastAnnotations.contains(annotation)
+            || handledEvent is! PointerHoverEvent
+            || latestEvent.position != handledEvent.position) {
+          if (annotation.onHover != null) {
+            annotation.onHover(latestEvent);
+          }
         }
       }
     }
   }
 
+  /// Mark all devices as dirty, and schedule a callback that is executed in the
```

Comment:
Is it still marking all of them dirty?

Reply:
Yes. After every new frame, all annotations might have been moved around, so it's necessary to check the state of all devices again.